### ISSUE TITLE
ERROR:  missing chunk number 0 for toast value 185589 in pg_toast_2619

### ISSUE THREAD

**Issue by Jasonysli** [postgres-x2/postgres-x2#165]
When running \d table, we got error msg:ERROR:  missing chunk number 0 for toast value 185589 in pg_toast_2619, it seems catalog has been corrupted!

foo=# select \* from pg_class where oid = 2619;
   relname    | relnamespace | reltype | reloftype | relowner | relam | relfilenode | reltablespace | relpages | reltuples | relallvisible | reltoastrelid | reltoastidxid | relhasindex | re
lisshared | relpersistence | relkind | relnatts | relchecks | relhasoids | relhaspkey | relhasrules | relhastriggers | relhassubclass | relispopulated | relfrozenxid | relminmxid |       re
lacl        | reloptions 
--------------+--------------+---------+-----------+----------+-------+-------------+---------------+----------+-----------+---------------+---------------+---------------+-------------+---
----------+----------------+---------+----------+-----------+------------+------------+-------------+----------------+----------------+----------------+--------------+------------+---------
------------+------------
 pg_statistic |           11 |   10818 |         0 |       10 |     0 |       12640 |             0 |       50 |       421 |            50 |          2840 |             0 | t           | f 
          | p              | r       |       26 |         0 | f          | f          | f           | f              | f              | t              |      7519316 |          1 | {pgxc=ar
wdDxt/pgxc} | 
(1 row)

Time: 1.322 ms
pg_statistic  got OID 2619.

**Comment by Jasonysli** [comment]
We have fixed the gixd fall back issue, the error still bothers us. I think may be it is caused by other issues.

**Comment by Jasonysli** [comment]
On another coordinator, when runing \d+,we got similar error msg:
foo=# \d+
ERROR:  catalog is missing 7 attribute(s) for relid 141685
foo=# select \* from pg_class where oid = 141685;
         relname          | relnamespace | reltype | reloftype | relowner | relam | relfilenode | reltablespace | relpages | reltuples | relallvisible | reltoastrelid | reltoastidxid | relh
asindex | relisshared | relpersistence | relkind | relnatts | relchecks | relhasoids | relhaspkey | relhasrules | relhastriggers | relhassubclass | relispopulated | relfrozenxid | relminmxi
d | relacl | reloptions 
--------------------------+--------------+---------+-----------+----------+-------+-------------+---------------+----------+-----------+---------------+---------------+---------------+-----
--------+-------------+----------------+---------+----------+-----------+------------+------------+-------------+----------------+----------------+----------------+--------------+----------
--+--------+------------
 v_query_test_table_1000w |        17446 |  141687 |         0 |    17736 |     0 |      141685 |             0 |        0 |         0 |             0 |             0 |             0 | f  
        | f           | p              | v       |        7 |         0 | f          | f          | t           | f              | f              | t              |            0 |  
0 |        | 
(1 row)

**Comment by Jasonysli** [comment]
The two error messages are both about catalog corrupted, so sad.

**Comment by koichi-szk** [comment]
Thanks for the report.    Does it happen with the master or REL1_2_STABLE?    Please test it with REL1_2_STABLE branch, not master because master does not have many fixes of PG itself.

Can you reproduce the error anytime or does it happen occasionally?   I found the same error in the past but I was not successful to reproduce it.

If you can, please let me know what branch you used.    I will write a patch to collect the info.    Please let me know the exact error message, if you have other one than you reported.

I can provide a patch this weekend or Monday for the latest.

Thank you;
—
Koichi

**Comment by koichi-szk** [comment]
Please find enclosed a patch to cause PANIC and get a core when the above error occurs.   You can take back trace for analysis.

Here it is:

```
diff --git a/src/backend/access/heap/tuptoaster.c b/src/backend/access/heap/tuptoaster.c
index 8781cb4..f3703e2 100644
--- a/src/backend/access/heap/tuptoaster.c
+++ b/src/backend/access/heap/tuptoaster.c
@@ -1732,7 +1732,7 @@ toast_fetch_datum(struct varlena * attr)
     * Final checks that we successfully fetched the datum
     */
    if (nextidx != numchunks)
-       elog(ERROR, "missing chunk number %d for toast value %u in %s",
+       elog(/* ERROR */ PANIC, "missing chunk number %d for toast value %u in %s",
             nextidx,
             toast_pointer.va_valueid,
             RelationGetRelationName(toastrel));
@@ -1956,7 +1956,7 @@ toast_fetch_datum_slice(struct varlena * attr, int32 sliceoffset, int32 length)
     * Final checks that we successfully fetched the datum
     */
    if (nextidx != (endchunk + 1))
-       elog(ERROR, "missing chunk number %d for toast value %u in %s",
+       elog(/* ERROR */ PANIC, "missing chunk number %d for toast value %u in %s",
             nextidx,
             toast_pointer.va_valueid,
             RelationGetRelationName(toastrel));
```

However, I suspect the cause is not here, there could be another module which destroyed some part of the storage.

I suspect there could be a buffer overflow.   Backtrace will help to see what catalog is damaged and what code we should investigate.    Could you test XC with valgrind?   This sometimes helps to find such bad bug.   We did it before 1.0 release and was helpful to find such potential bugs.

Sorry for asking so many things.   Hope this helps.

Thank you very much;

Thank you very much;

**Comment by koichi-szk** [comment]
Because XC depends upon palloc(), you may need to do something extra to check memory leak with valgrind.   I think there are some document available on the setup.    It might be in a file server of the office.

Let me check.   If I'm lucky, I can provide the setup info later on Monday.

Thank you;

**Comment by Jasonysli** [comment]
Hi koichi-szk, sorry for the delayed reply.I answer the question one by one:
1、OK, we will keep using REL1_2_STABLE code to do the test instead of master, in fact , we are using REL1_2_STABLE.
2、Our system was running with high work load, the issue happens on two of four coordinators. We did no special operation, just normal DDL and DML. I think maybe it was caused by catalog corruption, becasue we got "ERROR: catalog is missing 7 attribute(s) for relid 141685" msg at the same time.
3、In  pressure test, we keep the xc running on high workload, no memory leak found after 24 hours running. But we will keep observing memory info when running test.

**Comment by koichi-szk** [comment]
Thank you very much for the report.   Sorry I'm not successful to find a
guide how to use valgrind with XC.   Will let you know when found.

What is the relation with relied 141685?

I'm afraid there could be at least one buffer overflow, which runs very
rarely.   Will continue to review the code to find it.

Thank you again;

---

Koichi Suzuki

2015-06-01 11:53 GMT+09:00 Jasonysli notifications@github.com:

**Comment by Jasonysli** [comment]
Hi koichi-szk:
    Oid 141685 is a user defined relation, An ordinary table, no special. 
    What do you mean by saying buffer overflow? Do you mean shared buffer overflow?

**Comment by Jasonysli** [comment]
If it was caused by shared buffer overflow, may be we can help!

**Comment by koichi-szk** [comment]
I mean writing beyond palloc()-ed memory boundary.    This can destroy
palloc() control area, which can end up with destroying other independent
memory area.   This is not easy to discover this type of bug and valgrind
is very handy to detect this.

Regards;

---

Koichi Suzuki

2015-06-01 12:23 GMT+09:00 Jasonysli notifications@github.com:

**Comment by Jasonysli** [comment]
Ok, thanks for your infomation, we will try valgrind to see whether there is buffer over flow?

**Comment by koichi-szk** [comment]
Default valgrind detects malloc/free.   In our case, we have to detect
palloc/pfree instead.   Need to find options to tell this to valgrind.

https://bugzilla.redhat.com/show_bug.cgi?id=950585 can be a help.

Regards;

---

Koichi Suzuki

2015-06-01 12:29 GMT+09:00 Jasonysli notifications@github.com:

**Comment by Jasonysli** [comment]
Thank you, we will do the test next week!

**Comment by koichi-szk** [comment]
Hello;

Does anybody has updates to this issue?

**Comment by Jasonysli** [comment]
Sorry for the delayed reply. We used valgrind to verify whether there was buffer overflow last week, unfortunately, there is no such thing. Should  we focus on data inconsistance?

**Comment by koichi-szk** [comment]
Thanks.   When I looked into duplicate OIDs problem, I found some internal
XC-specific function's OID was duplicate with some toast-related functions.

This could have been the cause of the problem.   I will commit a fix for
this duplicate ASAP.   I hope you replace the binary with this new one,
reconstruct XC database cluster (database is not compatible, sorry) and do
the test to see if the same problem happens.

Hope this hits the point.

Thank you;

---

Koichi Suzuki

2015-07-01 11:55 GMT+09:00 Jasonysli notifications@github.com:
